extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -482,11 +488,124 @@ class LifecycleTask extends BackbeatTask {
+                    currentDate,
+                    transitions: rule.Transitions,
+                    objectLastModified: metadata.LastModified,
+                    objectStorageClass: metadata.StorageClass,
+                });
+                // Compare across rules to determine which is most applicable.
+                const ruleToApply = this._compareTransitionRules({
+                    currentDate,
+                    objectLastModified: metadata.LastModified,
+                    rule1: transitionRule,
+                    rule2: store.Transition,
+                });
+                store.Transition = ruleToApply;
+            }
+            // TODO: Add support for NoncurrentVersionTransitions.
             return store;
         }, {});
         /* eslint-enable no-param-reassign */
     }
 
+    /**
+     * Get the Unix timestamp of the given date.
+     * @param {string} date - The date string to convert to a Unix timestamp
+     * @return {number} - The Unix timestamp
+     */
+    _getTimestamp(date) {
+        return new Date(date).getTime();
+    }
+
+    /**
+     * Find the Unix time at which the transition should occur.
+     * @param {object} transition - A transition from the lifecycle transitions
+     * @param {string} objectLastModified - The object's last modified date
+     * @return {number|undefined} - The normalized transition timestamp
+     */
+    _getTransitionTimestamp(transition, objectLastModified) {
+        if (transition.Date !== undefined) {
+            return this._getTimestamp(transition.Date);
+        }
+        if (transition.Days !== undefined) {
+            const lastModifiedTime = this._getTimestamp(objectLastModified);
+            const oneDay = 24 * 60 * 60 * 1000; // Milliseconds in a day.
+            return lastModifiedTime + (transition.Days * oneDay);
+        }
+        return undefined;
+    }
+
+    /**
+     * Find the most relevant trantition rule for the given transitions array.
+     * @param {object} params - The function parameters
+     * @param {array} params.transitions - Array of lifecycle rule transitions
+     * @param {string} params.currentDate - The current date to compare against
+     * @param {string} params.objectLastModified - The object's last modified
+     * date
+     * @param {string} params.objectStorageClass - The object's AMZ storage
+     * class
+     * @return {object} The most applicable transition rule
+     */
+    _getTransitionRule(params) {
+        const { transitions, objectLastModified, currentDate } = params;
+        let result;
+        transitions.forEach(transition => {
+            const shouldTransition =
+                this._getTimestamp(currentDate) >=
+                this._getTransitionTimestamp(transition, objectLastModified);
+            if (shouldTransition === false) {
+                return undefined;
+            }
+            // Is this transition rule the most recent one in the transitions?
```

Comment:
wouldn't we want the transition rule that is valid and oldest? Why most recent one?

Reply:
That's right, and a more precise way of wording this since "recent" can be interpreted in different ways. I'll update accordingly.

Reply:
ah sorry, my misunderstanding, thanks for clarifying 👍

Reply:
No misunderstanding—the comment should be written such that it cannot be misinterpreted. Thanks for bringing it up! 😸